Init python/django/lint/ci

twidi opened the pull request:
Python and django scaffolding with linters (missing doc about this) + django models following the ADRs, and CircleCI workflows config (see https://circleci.com/workflow-run/607fe31b-80db-4321-b4ac-6a1517666c2d)

sampaccoud commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+        verbose_name_plural: str = _("playlists")
+
+
+class PlaylistVideo(BaseModel):
```
Why not just Playlist? 
If you want to add video it should be VideoPlaylist in english?

twidi replied:
It represents a video a in a playlist. Not the video or the playlist itself. And then used as the through table between these two models.

So for me, `VideoPlaylist` or `PlaylistVideo` are the same.

sampaccoud commented on marsha/settings.py:
```diff
@@ -0,0 +1,99 @@
+"""Django settings for marsha project.
```
I think @jmaupetit will propose to use DjangoConfiguration.

jmaupetit replied:
:wink: you read my mind.

twidi replied:
If it's your choice I'll follow you. Not fan of this kind of heavy machinery that totally changes how settings are managed. But it's only my opinion.

sampaccoud replied:
I think for coherence in our projects it would be nice to use it.

twidi replied:
ok then

twidi replied:
I will but:

- non standard way of defining settings
- cannot define a settings based on a previous one loaded from the env
- cannot use django-admin.py

Digging into the documentation and the code means I won't use it for any project, but I follow your recommandations

twidi commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            - linter_flake8
+            - linter_pylint
+            - check_migrations
+            - django_checks
```
Checking if a package can be built for a PR is a good thing. If not, the PR is bad.

A next step could check if the package could be installed.

sampaccoud replied:
:+1: 
... and publish it publically or to our gemfury upon taggging right?

twidi replied:
I really want to built it every time. But only publish for tags, ok?

sampaccoud replied:
Yes let's do that!

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+          name: Check migrations
+          command: |
+            source ~/venv/bin/activate
+            django-admin.py makemigrations --check --dry-run
```
:1st_place_medal:

sampaccoud replied:
I didn't comment here but I'm not really sure it will allow us doing upgrades as we want to do them ie code base version N and N+1 both work with schema S and S+1 (this involves voluntarily delaying migrations and thus desynchronizing them from the code). But we can see this later, maybe I'm wrong.

twidi replied:
I propose to think about it later. I need to know more about your way to do this and also think about how to not miss migrations.

jmaupetit replied:
I might be wrong, but I think this will work even with our multi-steps migration strategy.

sampaccoud replied:
Yes leave it it's a good idea to check it and we'll see whaf happens with deployments.

twidi commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+        unique_together: UniqueTogether = [("video", "playlist")]
+
+
+class PlaylistRight(BaseModel):
```
In fact it was my first choice... so, OK :)

twidi commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+
+    class Meta:
+        """Options for the ``SiteAdmin`` model."""
+
```
I was hesitating on this. But yes ok, good idea

sampaccoud replied:
... so should we keep the name of the app in the table names? I thought the idea of naming them ourselves was to name them differently and remove the link to app names.

twidi replied:
It was in my mind. No need for `core_`, even `marsha_` prefixes.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+      - run:
+          name: Build package
+          command: |
+            sed -i -e "s/^version = .*$/version = $(git describe --tags)/" setup.cfg
```
IMO this package build should be triggered upon a new git tag pushed to the repository. Hence we suppose, the `setup.cfg` file is already up-to-date.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            sed -i -e "s/^version = .*$/version = $(git describe --tags)/" setup.cfg
+            source ~/venv/bin/activate
+            python setup.py sdist
+            git checkout -- setup.cfg
```
You do not need to reset the repository state. When the job ends, the container is trashed.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            - linter_flake8
+            - linter_pylint
+            - check_migrations
+            - django_checks
```
You should add a filter here to build package only for new releases. See this example: https://github.com/openfun/richie/blob/master/.circleci/config.yml#L559

twidi replied:
Did you read my comment above? I really want to check that the whole code is correct. So trying to build the package is a good thing. If I only try to build it when the code is finally merged and tagged and the build doesn't work, we are in a wrong state

jmaupetit replied:
Okay then. Remember that we had concurrent reviews with @sampaccoud.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            - linter_flake8
+            - linter_pylint
+            - check_migrations
+            - django_checks
```
Same remark for the documentation.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            python setup.py sdist
+            git checkout -- setup.cfg
+      - store_artifacts:
+          path: dist/
```
Maybe that's a first approach, but do you think we will use those artifacts? Why not pushing the new package to a PyPI repository?

twidi replied:
For build: next step could be to create a venv and install the package, for example. And doing it manually at some point could also be a simple option.
For doc: I can read the doc. I thought of creating a version in the github pages for the project but it's not possible with the way github pages are managed by github without polluting the branches (if it was another repository it would have been ok)

Is there any cost tied to storing artifacts? Are they trashed after some time?

jmaupetit replied:
No there is no cost associated with those artifacts. And they are automatically trashed after a few days.

twidi replied:
ok so we can keep them to be usable in case of need.

jmaupetit replied:
Yes.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+            sphinx-apidoc -f -o source/ ../marsha/
+            make html
+      - store_artifacts:
+          path: docs/_build
```
Same remark as for the package.

jmaupetit commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+        unique_together: UniqueTogether = [("video", "playlist")]
+
+
+class PlaylistRight(BaseModel):
```
Why not naming it `UserPlaylist` or something similar? Not sure `FooRight` is the way to go.

twidi replied:
As proposed by sam, `PlaylistAccess` seems good to me

jmaupetit commented on marsha/settings.py:
```diff
@@ -0,0 +1,99 @@
+"""
+
+
+import environ
```
Do you know `django-configurations`[1]? I particularly like the class-base approach to define settings. And they also support 12 factors app derived settings.

[1] https://github.com/jazzband/django-configurations

twidi replied:
As I said in my answer to sam:

> If it's your choice I'll follow you. Not fan of this kind of heavy machinery that totally changes how settings are managed. But it's only my opinion.

twidi replied:
ok sam decided to go with this heavy machinery ;)

jmaupetit commented:
Sorry for duplicated questions in my review. I did not expect that @sampaccoud was also reviewing your code.

twidi commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+        to=Organization,
+        related_name="playlists",
+        # don't delete a playlist if its organization is deleted
+        on_delete=models.SET_NULL,
```
Soft deleting adds a layer of complexity (managers, base model...). so I propose to add this in a next PR, keeping things liike this for now. What do you think?

sampaccoud replied:
Agree to do soft delete in an other PR but corollary is maybe to use RESTRICT on all these FK to avoid losing information in the meantime?

twidi replied:
"In the meantime"... we are not in prod yet. But ok ;)

twidi replied:
it's `PROTECT` in Django, btw

twidi commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+        verbose_name_plural: str = _("videos")
+
+
+class VideoTrackBase(BaseModel):
```
About the name. I like ends base classes with `Base`, like I do with `Mixin` for mixins. So `TrackBase`. So I agree to remove `Video`, renaming `VideoTrackBase` to `TrackBase`.

About `VideoTrack`, in #3 I talked about possibilities to have other kinds of video, asking for discussion about his.

> At this time we cannot think of another kind of attached video, so it's called ``SignsTrack``. If it happen in the future, it could be renamed ``AuxiliaryTrack``, with a ``kind`` field.

The ADR was accepted like this so... 

Do you think of any other kind of video that could be attached to a main video?

sampaccoud replied:
I can't think of other kinds of video yet... so... ok to keep SignTrack.

For the Base, I have the feeling that everywhere Base is used as a prefix not a suffix. So I think we should follow this convention. First example is BaseModel from which you inherit here ;-)
Another example: https://github.com/divio/django-cms/blob/develop/cms/extensions/models.py
I'm sure there are many examples in Django's code base itself.

Also I feel it is clearer that `Base` is at the same place as what replaces it in children classes:
BaseTrack become (Video|Audio|Sign)Track.

sampaccoud replied:
If you can't agree on this, maybe submit it to the vote on Slack ;-)

twidi replied:
Ok you convinced me!

twidi commented on setup.cfg:
```diff
@@ -0,0 +1,139 @@
+[pycodestyle]
+max-line-length = 99
+
+
```
because black doesn't manage `.cfg` files and I made a huge mistake here :p

jmaupetit replied:
:laughing:

twidi commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+        - .
+  restore_repo: &restore_repo
+    restore_cache:
+      key: v1-repo-{{ .Branch }}-{{ .Revision }}
```
I really don't want to do a new checkout at the beginning of each jobs. I can get that it may be faster, but I see it like "steps". First checkout. Then use this base to install code. Then use this new base to do the linters, etc...

I think we talk in durations less than a second or two here so we can ignore this at least for now.

Do you agree?

twidi replied:
I got your concerns. Note that I got this way of doing things from the... circle-ci frontend repository itself !! https://github.com/circleci/frontend/blob/master/.circleci/config.yml

About minimizing side effects, you get it wrong. There is no side effects because the directory is cached just after being checkouted / pip installe. And freshly restored as is each time.

So the only thing I agree with you is to not share venv betweens differents push-jobs (but keeping it between steps of a same push because of what I said above)

jmaupetit replied:
Fair enough.

twidi commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+    <<: *python_only_config
+    steps:
+      - checkout
+      - *save_repo
```
I guess it's related to your other comment about speed?

twidi commented on marsha/core/models.py:
```diff
@@ -0,0 +1,392 @@
+    )
+    video_language: str = models.CharField(
+        max_length=5,
+        verbose_name=_("video language"),
```
yes it should. Do you have a list for 'en_US' kind of languages to propose?

sampaccoud replied:
use settings.LANGUAGES and set it to just english and french for a start?

twidi replied:
ok

twidi commented:
@sampaccoud @jmaupetit

I pushed a few commits for changes not needing further discussions.

But still a lots of things waiting us to agree about

jmaupetit commented on .circleci/config.yml:
```diff
@@ -0,0 +1,235 @@
+      keys:
+        - v2-venv-{{ .Branch }}-{{ .Revision }}
+        - v2-venv-{{ .Branch }}
+        - v2-venv
```
> In fact in term of speed, this is a real improvement, and in fact the biggest one.

Indeed. But that's not the point :wink: 

We need to ensure that installing dependencies as defined in a version v is working without side effects of the version v-1's environment. By doing so, your build step is not reproducible and you cannot be sure that a fresh virtualenv is functional.

> Think upgrading a production server with a new version of a package: you don't trash existing installed things, you just apply the upgrade.

I don't think production server upgrade as you do :wink:: we will trash a running container and replace it by a fresh one build from scratch without prior history. Think reproducibility.

twidi replied:
Yes of course, in your way/recent way of doing deployment. But for years, to upgrade a production server we didn't trash everything ;)

I agreed on another comment to not share venv between different push-jobs.

jmaupetit commented on .circleci/config.yml:
```diff
@@ -77,31 +51,37 @@ jobs:
+          root: "~/"
+          paths:
+            - venv
+            - marsha
```
I've never used workspaces, this is a brilliant idea! From what I've read [1], this will also speed up the build as it's faster than using the cache. :+1: 

[1] https://hackernoon.com/circleci-performance-difference-between-cache-and-workspace-5567679c3601

twidi replied:
Yes it's the point and it is exactly what I was looking for. Also, it is only update for next jobs if `persist_to_workspace` is used, it's why I do not persist it after lint jobs.

twidi commented:
@sampaccoud @jmaupetit

I finished all we talked about, tweaked a lot of things, added a makefile, added docs about development, configured readthedocs for this, and squashed all the fixup commits into the original ones.

I think I'm good on this PR now!

twidi commented:
PS: doc is here http://marsha.readthedocs.io/en/feature-twidi-init-django/

jmaupetit commented on Makefile:
```diff
@@ -0,0 +1,75 @@
+.PHONY: full-clean
+full-clean: clean
+	@echo "$(BOLD)Full cleaning$(RESET)"
+	@rm -rf .mypy_cache
```
Can you add a help target (see [1]) and a description for each target to generate the help:

```Makefile
.PHONY: doc
doc: ## generate project's HTML documentation
	@echo "$(BOLD)Building documentation$(RESET)"
	@cd docs && $(MAKE) html
```

[1] https://github.com/openfun/richie/blob/master/Makefile#L105

jmaupetit commented on docs/dev.rst:
```diff
@@ -0,0 +1,315 @@
+- `flake8-mypy <https://pypi.org/project/flake8-mypy/>`_ to check typing inconsistencies via ``mypy`` (see :ref:`section-mypy`).
+- `flake8-docstrings <https://pypi.org/project/flake8-docstrings/>`_ to check docstrings (see :ref:`section-docstrings`)
+
+To check your code, simply run the command:
```
Some say that some words such as "_simple_" or "_simply_" must be avoided in a technical documentation as it may be offending to some people working hard to understand what's not so obvious.

jmaupetit commented on docs/dev.rst:
```diff
@@ -0,0 +1,315 @@
+
+The type expected for a ``OneToOneField`` is the pointed model.
+
+And on the pointed model, we set the type the related name, to be the source model.
```
I think I get what you meant, but this needs to be re-phrased IMO.

jmaupetit commented on docs/dev.rst:
```diff
@@ -0,0 +1,315 @@
+    class Foo(models.Model):
+        # we tell mypy that the attribute ``bar`` is of type ``Bar``
+        bar: "Bar" = models.OneToOneField(
+        to=Bar, related_name="the_foo")
```
Indentation issue?

jmaupetit commented on docs/dev.rst:
```diff
@@ -0,0 +1,315 @@
+
+The type expected for a ``ForeignKey`` is the pointed model.
+
+On the pointed model, we have a many-to-one relation. We use a type specifically defined for that, ``ReverseFKType``,
```
`s/relation/relationship/`?

jmaupetit commented on marsha/settings.py:
```diff
@@ -0,0 +1,102 @@
+
+class Dev(Configuration):
+    """Django settings in Dev mode."""
+
```
I usually define a `Base` configuration to inherit from. See Richie's example [1].

[1] https://github.com/openfun/richie/blob/master/richie/settings.py#L16

twidi commented:
@jmaupetit: I added 3 fixup commits regarding your remarks, and a new doc about environment variables. We'll have later to better organize doc, but not sure now how to do it, without anything more to put in it :)

jmaupetit commented on docs/dev.rst:
```diff
@@ -144,8 +144,8 @@ Foreign keys
 
-On the pointed model, we have a many-to-one relation. We use a type specifically defined for that, ``ReverseFKType``,
-defined in ``marsha.stubs``.
+On the pointed model we have a many-to-one relationshop.
```
`s/relationshop/relationship/`

jmaupetit commented:
Go, go, go!
